lib/upipe-modules/upipe_audio_merge.c

```diff
@@ -0,0 +1,638 @@
+        upipe_audio_merge_from_sub_mgr(upipe->mgr);
+    struct upipe_audio_merge_sub *upipe_audio_merge_sub =
+        upipe_audio_merge_sub_from_upipe(upipe);
+
+    if (flow_def == NULL)
+        return UBASE_ERR_INVALID;
+
+    /* reject interleaved audio */
+    uint8_t channels = 0, planes = 0;
+    UBASE_ERROR(upipe, uref_sound_flow_get_planes(flow_def, &planes));
+    UBASE_ERROR(upipe, uref_sound_flow_get_channels(flow_def, &channels));
+    if (planes != channels) {
+        upipe_err(upipe, "Interleaved audio not supported");
+        return UBASE_ERR_INVALID;
+    }
+    /* compare against stored flowdef (if we have one) and reject if formats don't match */
+    if (upipe_audio_merge->sub_flow_def
+            && !upipe_audio_merge_match_flowdefs(flow_def, upipe_audio_merge->sub_flow_def))
+    {
+        upipe_err(upipe, "Subpipe has non-matching audio format!");
+        return UBASE_ERR_INVALID;
+    }
+
+    /* .. and if we don't have one store this one */
+    if (!upipe_audio_merge->sub_flow_def)
+        upipe_audio_merge->sub_flow_def = uref_dup(flow_def);
+
+    /* store in the subpipe structure itself for later reference */
+    uref_free(upipe_audio_merge_sub->flow_def);
+    upipe_audio_merge_sub->flow_def = uref_dup(flow_def);
+
+    return UBASE_ERR_NONE;
+}
+
+/** @internal @This allocates an input subpipe of an audio_merge pipe.
+ *
+ * @param mgr common management structure
+ * @param uprobe structure used to raise events
+ * @param signature signature of the pipe allocator
+ * @param args optional arguments
+ * @return pointer to upipe or NULL in case of allocation error
+ */
+static struct upipe *upipe_audio_merge_sub_alloc(struct upipe_mgr *mgr,
+                                                 struct uprobe *uprobe,
+                                                 uint32_t signature, va_list args)
+{
+    struct upipe *upipe = upipe_audio_merge_sub_alloc_void(mgr,
+                            uprobe, signature, args);
+    if (unlikely(upipe == NULL)) {
+        return NULL;
+    }
+
+    struct upipe_audio_merge_sub *upipe_audio_merge_sub =
+                            upipe_audio_merge_sub_from_upipe(upipe);
+
+    upipe_audio_merge_sub_init_urefcount(upipe);
+    upipe_audio_merge_sub_init_sub(upipe);
+    upipe_audio_merge_sub->uref = NULL;
+    upipe_audio_merge_sub->flow_def = NULL;
+    upipe_throw_ready(upipe);
+    return upipe;
+}
+
+/** @internal @This processes control commands on an input subpipe of an
+ *  audio_merge pipe.
+ *
+ * @param upipe description structure of the pipe
+ * @param command type of command to process
+ * @param args arguments of the command
+ * @return an error code
+ */
+static int upipe_audio_merge_sub_control(struct upipe *upipe,
+                                         int command, va_list args)
+{
+    UBASE_HANDLED_RETURN(upipe_audio_merge_sub_control_super(upipe, command, args));
+    switch (command) {
+        case UPIPE_REGISTER_REQUEST: {
```

Comment:
You can use the helper `UBASE_HANDLED_RETURN(upipe_control_provide_request(upipe, command, args));`